### PR TITLE
feat(examples): add blob support to the rollup ExEx

### LINKED ISSUE #7558: EIP4844 blob sidecars ExEx example
Blocked by https://github.com/paradigmxyz/reth/issues/7419

An example of ExEx that queries local blob pool and asks the CL for blob sidecars of transactions from certain senders (e.g. rollup batch submitter like https://etherscan.io/address/0x6887246668a3b87f54deb3b94ba47a6f63f32985)

### PR DESCRIPTION
**Pull request by shekhirin**
Closes https://github.com/paradigmxyz/reth/issues/7558

This PR adds support for rollup block submissions using either calldata or EIP-4844 blobs https://github.com/paradigmxyz/reth/blob/b5ba0c23661c9a26023a5c3b86fb08ac376a40d7/examples/exex/rollup/src/execution.rs#L178-L220 

As well as an additional check for the `blockDataHash` commitment https://github.com/paradigmxyz/reth/blob/b5ba0c23661c9a26023a5c3b86fb08ac376a40d7/examples/exex/rollup/src/execution.rs#L222-L225

---

The diff is quite large due to moving decoding- and execution-related code to a separate file.

### CODE REVIEW

File: examples/exex/rollup/src/execution.rs [hunk 1]
```diff
@@ -0,0 +1,461 @@
...
+    block_data: Bytes,
+    block_data_hash: B256,
+) -> eyre::Result<Vec<(TransactionSigned, Address)>> {
+    // Get raw transactions either from the blobs, or directly from the block data
+    let raw_transactions = if matches!(tx.tx_type(), TxType::Eip4844) {
+        let blobs: Vec<_> = if let Some(sidecar) = pool.get_blob(tx.hash)? {
+            // Try to get blobs from the transaction pool
+            sidecar.blobs.into_iter().zip(sidecar.commitments).collect()
+        } else {
+            // If transaction is not found in the pool, try to get blobs from the Blobscan
```
Comment (onbjerg) [inline examples/exex/rollup/src/execution.rs:184]:
```suggestion
            // If transaction is not found in the pool, try to get blobs from Blobscan
```

File: examples/exex/rollup/src/execution.rs [hunk 2]
```diff
@@ -0,0 +1,461 @@
...
+    db::{states::bundle_state::BundleRetention, BundleState},
+    DatabaseCommit, StateBuilder,
+};
+use reth_tracing::tracing::debug;
+
+use crate::{db::Database, RollupContract, CHAIN_ID, CHAIN_SPEC};
+
+/// Execute a rollup block and return (block with recovered senders)[BlockWithSenders], (bundle
+/// state)[BundleState] and list of (receipts)[Receipt].
+pub async fn execute_block<Pool: TransactionPool>(
```
Comment (onbjerg) [inline examples/exex/rollup/src/execution.rs:27]: this is growing quite large, maybe we can extract some of the evm setup or prechecks into another function?

**Review by onbjerg — APPROVED** [review]
lgtm otherwise

File: examples/exex/rollup/src/execution.rs (hunk 2 above)
Comment (shekhirin) [reply examples/exex/rollup/src/execution.rs:27]: split into more functions

File: examples/exex/rollup/src/execution.rs [hunk 3]
```diff
@@ -0,0 +1,489 @@
...
+    block_data: Bytes,
+    block_data_hash: B256,
+) -> eyre::Result<Vec<(TransactionSigned, Address)>> {
+    // Get raw transactions either from the blobs, or directly from the block data
+    let raw_transactions = if matches!(tx.tx_type(), TxType::Eip4844) {
+        let blobs: Vec<_> = if let Some(sidecar) = pool.get_blob(tx.hash)? {
+            // Try to get blobs from the transaction pool
+            sidecar.blobs.into_iter().zip(sidecar.commitments).collect()
+        } else {
+            // If transaction is not found in the pool, try to get blobs from Blobscan
```
Comment (gakonst) [inline examples/exex/rollup/src/execution.rs:139]: what is the right api for pulling blobs from CL? Should we open issue for this?

File: examples/exex/rollup/src/main.rs [hunk 4]
```diff
@@ -96,12 +92,21 @@ impl<Node: FullNodeComponents> Rollup<Node> {
                         ..
                     }) = call
                     {
-                        match execute_block(&mut self.db, &header, blockData) {
+                        match execute_block(
```
[the remark this answers is not in the document]
Comment (shekhirin) [reply examples/exex/rollup/src/main.rs:95]: ideally, yes – but we're not ready with E2E testing for the ExExes yet. The perfect workflow is a mock `ExExContext` that you can init your ExEx with, and then test the whole thing E2E.

File: examples/exex/rollup/src/execution.rs [hunk 5]
```diff
@@ -0,0 +1,461 @@
...
+        timestamp: u64::try_from(header.confirmBy)?,
+        base_fee_per_gas: Some(base_fee_per_gas),
+        ..Default::default()
+    };
+
+    // Decode transactions
+    let transactions = decode_transactions(pool, tx, block_data, block_data_hash).await?;
+
+    // Configure EVM
+    let evm_config = EthEvmConfig::default();
```
[the remark this answers is not in the document]
Comment (shekhirin) [reply examples/exex/rollup/src/execution.rs]: we'll be able to use `EthExecutorProvider` when it supports custom errors and not just `ProviderError`, @mattsse is aware of this

File: examples/exex/rollup/src/execution.rs (hunk 3 above)
Comment (mattsse) [reply examples/exex/rollup/src/execution.rs:139]: coming here #7443

**Review by mattsse — APPROVED** [review]

File: examples/exex/rollup/src/main.rs [hunk 6]
```diff
@@ -162,6 +172,7 @@ impl<Node: FullNodeComponents> Rollup<Node> {
         events.reverse();
 
         for (_, tx, event) in events {
+            let tx_hash = tx.recalculate_hash();
```
[the remark this answers is not in the document]
Comment (shekhirin) [reply examples/exex/rollup/src/main.rs]: it's expensive and needed only for logs, fixed

**Pull request merged**